server/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java

```diff
@@ -63,18 +59,41 @@
     public static final String NAME = "terms";
 
     private final String fieldName;
-    private final List<?> values;
     private final TermsLookup termsLookup;
     private final Supplier<List<?>> supplier;
 
+    /**
+     * Store terms as a {@link BytesRef}.
+     * <p>
+     * When users send a query contain a lot of terms, A {@link BytesRef} can help
+     * gc and reduce the cost of {@link #doWriteTo}, which can be quite slow for many
+     * terms.
+     */
+    private final BytesRef valueRef;
+
+    /**
+     * This is for lower version requests compatible.
+     * <p>
+     * If we do not keep this, it could be expensive when receiving a request from
+     * lower version.
+     * We have to read the value list by {@link StreamInput#readGenericValue},
+     * serialize it into {@link #valueRef}, and then deserialize it again when
+     * {@link #doToQuery} called}.
+     * <p>
+     * ToDo: remove in 9.0.0
+     */
+    private final List<?> values;
+    private final boolean stale; //indicate if this comes from a lower version
```

Comment:
@jimczi  Thank you for the review, I think this idea is really excellent! i've made some effort to optimize the code based on this idea in a new [commit](https://github.com/elastic/elasticsearch/pull/67223/commits/b374ec734abe868276853078679073bec602ed5c). Here are some notes about this modification
1. Using a `Iterable` to lazily load data from `BytesRef` is very good, but a `MappedFieldType#termsQuery` used to expect a `list` of terms, which means we still had to convert the `Iterable` to `List` to get a TermsQuery. This can make the lazily load meaningless, so i changed the method signature from a `List` to a `Collection`.
2. The reason that I did not convert `List` to `Iterable` but `Collection` is that we usually use a `List#size()` to pre-construct another Collection, e.g. a `LongHashSet` in `DateScriptFieldType`. This can help when we add data into the new collection. we can not get a `size` value from `Iterable`, but we can get it from `Collection`.